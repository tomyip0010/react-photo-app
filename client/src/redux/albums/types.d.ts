type AlbumType = {
  id: number,
  title: string,
  userId: number,
}

interface AlbumsStoreType {
  isFetching: boolean,
  fetchSuccess: boolean,
  albumList: Array<AlbumType>,
  totalCount: number,
  filter: FilterType,
  error?: any,
}