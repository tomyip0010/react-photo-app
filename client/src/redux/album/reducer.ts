import { FETCH_ALBUM_PHOTOS } from './action';

const initialState: AlbumStoreType = {
  isFetching: false,
  fetchSuccess: false,
  albumPhotos: [],
  filter: {},
  error: {},
};

export default function reducer(
  state = initialState,
  action: GenericActionType,
) {
  switch (action.type) {
    case FETCH_ALBUM_PHOTOS.REQUEST:
      return {
        ...state,
        albumPhotos: [],
        isFetching: true,
      };
    case FETCH_ALBUM_PHOTOS.SUCCESS:
      return {
        ...state,
        albumPhotos: action.result.data,
        isFetching: false,
        fetchSuccess: true,
      };
    case FETCH_ALBUM_PHOTOS.FAILURE:
      return {
        ...state,
        isFetching: false,
        fetchSuccess: false,
        error: action.error,
      };
    default:
      return initialState;
  }
}