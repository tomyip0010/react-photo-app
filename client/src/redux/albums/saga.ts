import {
  call, all, takeLatest, put, select,
} from 'redux-saga/effects';
import { FETCH_ALBUM_LIST } from './action';
import { apiRequest } from 'helper/apiClient';

function* watchfetchAlbumPhoto(data: GenericActionType) {
  const { params } = data;
  if (isNaN(params.offset) || isNaN(params.limit)) {
    const filter = yield select((state: ReduxStoreType) => state.albums.filter);
    if (!params.offset) {
      params.offset = filter.offset;
    }
    if (!params.limit) {
      params.limit = filter.limit;
    }
  }
  const response = yield call(apiRequest, 'get', '/albums', data);

  if (response) {
    yield put({
      type: FETCH_ALBUM_LIST.SUCCESS,
      result: response,
    });
  }
  // if (error) {
  //   console.log('>>>>>>with error', error);
  //   // yield call(handleError, FETCH_ALBUM_LIST, error);
  // } else if (response) {
  //   yield put({
  //     type: FETCH_ALBUM_LIST.SUCCESS,
  //     result: response,
  //   });
  // }
}

export default function* albumsSaga() {
  yield all([
    takeLatest(FETCH_ALBUM_LIST.REQUEST, watchfetchAlbumPhoto),
  ]);
}
