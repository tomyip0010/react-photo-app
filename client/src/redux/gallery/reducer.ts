import { FETCH_ALBUM_LIST } from './action';

const initialState: GalleryStoreType = {
  isFetching: false,
  fetchSuccess: false,
  albumList: [],
  totalCount: 0,
  filter: {
    limit: 20,
    offset: 0,
  },
  error: {},
};

export default function reducer(
  state = initialState,
  action: GenericActionType,
) {
  switch (action.type) {
    case FETCH_ALBUM_LIST.REQUEST:
      return {
        ...state,
        albumList: [],
        isFetching: true,
      };
    case FETCH_ALBUM_LIST.SUCCESS:
      return {
        ...state,
        albumList: action.result.data,
        totalCount: action.result.totalCount,
        filter: {
          limit: action.result.limit,
          offset: action.result.offset,
        },
        isFetching: false,
        fetchSuccess: true,
      };
    case FETCH_ALBUM_LIST.FAILURE:
      return {
        ...state,
        isFetching: false,
        fetchSuccess: false,
        error: action.error,
      };
    default:
      return initialState;
  }
}