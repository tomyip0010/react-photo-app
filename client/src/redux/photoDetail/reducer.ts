import { FETCH_PHOTO_DETAIL } from './action';

const initialState: PhotoDetailStoreType = {
  isFetching: false,
  fetchSuccess: false,
  photoDetail: {},
  error: {},
};

export default function reducer(
  state = initialState,
  action: GenericActionType,
) {
  switch (action.type) {
    case FETCH_PHOTO_DETAIL.REQUEST:
      return {
        ...state,
        photoDetail: {},
        isFetching: true,
      };
    case FETCH_PHOTO_DETAIL.SUCCESS:
      return {
        ...state,
        photoDetail: action.result.data,
        isFetching: false,
        fetchSuccess: true,
      };
    case FETCH_PHOTO_DETAIL.FAILURE:
      return {
        ...state,
        isFetching: false,
        fetchSuccess: false,
        error: action.error,
      };
    default:
      return initialState;
  }
}