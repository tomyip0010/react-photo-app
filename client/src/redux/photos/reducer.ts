import { FETCH_ALBUM_PHOTOS, CLEAR_FILTER } from './action';

const initialState: PhotosStoreType = {
  isFetching: false,
  fetchSuccess: false,
  albumPhotos: [],
  totalCount: 0,
  filter: {
    limit: 20,
    offset: 0,
  },
  error: {},
};

export default function reducer(
  state = initialState,
  action: GenericActionType,
) {
  switch (action.type) {
    case FETCH_ALBUM_PHOTOS.REQUEST:
      return {
        ...state,
        albumPhotos: [],
        isFetching: true,
      };
    case FETCH_ALBUM_PHOTOS.SUCCESS:
      return {
        ...state,
        albumPhotos: action.result.data,
        totalCount: action.result.totalCount,
        filter: {
          limit: action.result.limit,
          offset: action.result.offset,
        },
        isFetching: false,
        fetchSuccess: true,
      };
    case FETCH_ALBUM_PHOTOS.FAILURE:
      return {
        ...state,
        isFetching: false,
        fetchSuccess: false,
        error: action.error,
      };
    case CLEAR_FILTER:
      return {
        ...state,
        filter: {
          limit: 20,
          offset: 0,
        },
      }
    default:
      return state;
  }
}