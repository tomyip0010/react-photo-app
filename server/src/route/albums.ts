import { Request, Response, NextFunction } from 'express';
import { apiRequest } from '../helper/apiClient';
import { getRedisKey, exSetRedisKey } from '../helper/cacheManager';
import HttpException from '../helper/HttpException';

interface AlbumsType {
  userId: number,
  id: number,
  title: string,
}

interface ResponseType {
  success: boolean,
  offset: number,
  limit: number,
  totalCount: number,
  data: Array<AlbumsType>
}

async function getAlbumContent(req: Request, res: Response, next: NextFunction) {
  const refresh = req.query.refresh === 'true';
  const offset = req.query.offset && !isNaN(req.query.offset) ? Number(req.query.offset) : 0;
  const limit = req.query.limit && !isNaN(req.query.limit) ? Number(req.query.limit) : 20;
  const key = 'albums-content';
  const response: ResponseType = {
    success: false,
    offset,
    limit,
    totalCount: 0,
    data: [],
  };

  try {
    let dataContent;
    if (!refresh) {
      const cachedContent = await getRedisKey(key);
      if (cachedContent) {
        console.log(`${key}, with content from cache `);
        dataContent = JSON.parse(cachedContent);
      }
    }
    if (refresh || !dataContent) {
      console.log(`${key}, without cahce, fetching from api`);
      dataContent = await apiRequest('get', '/albums', {});
    }
    if (dataContent) {
      const startPos = offset;
      const endPos = offset + limit;
      response.data = dataContent.slice(startPos, endPos);
      response.success = true;
      response.totalCount = dataContent.length;
      exSetRedisKey(key, JSON.stringify(dataContent), 86400);
      return res.send({ response });
    }
    return res.send({ response });
  } catch (err) {
    console.log('fetch albums content error ', err);
    let errorCode = 500;
    let errorMessages = 'Something went wrong!';
    if (err.response) {
      errorCode = err.response.status;
      errorMessages = err.response.statusText;
    }
    next(new HttpException(errorCode, errorMessages));
  }
}

export default getAlbumContent;
