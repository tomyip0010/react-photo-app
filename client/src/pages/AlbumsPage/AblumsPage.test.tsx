import * as React from 'react';
import { create } from 'react-test-renderer';
import { initialState } from 'redux/albums/reducer';
import { mount } from 'enzyme';
import { AlbumsPage } from './index';

describe("AlbumsPage", () => {
  let props: any;
  beforeEach(() => {
    props = {
      ...initialState,
      location: {
        search: ''
      },
      history: '',
      fetchAlbumList: jest.fn(),
    }
  });

  test("render without fail and match snapshot", () => {
    const page = create(<AlbumsPage {...props} />);
    expect(page.toJSON()).toMatchSnapshot();
  });

  test("fetch album in did mount", () => {
    const wrapper = mount(<AlbumsPage {...props} />);
    expect(props.fetchAlbumList).toHaveBeenCalled();
  });

  // test("render loading when is fetching data", () => {

  // });
});