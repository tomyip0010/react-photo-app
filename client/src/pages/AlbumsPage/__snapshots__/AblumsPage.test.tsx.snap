// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlbumsPage Matches the snapshot 1`] = `
<div
  className="albumsPage"
>
  <div>
    <nav
      className="navbar navbar-expand navbar-dark bg-dark"
    >
      <a
        className="sc-bdVaJa heUcHr navbar-brand"
        href="#home"
      >
        <img
          alt="logo"
          src="logo.svg"
        />
      </a>
      <div
        className="sc-bwzfXH eocOaE"
      >
        React Photo App
      </div>
    </nav>
    <div
      className="sc-htpNat jmLCxE"
    >
      <div
        className="sc-ifAKCX itiWWq"
      >
        Albums
      </div>
    </div>
  </div>
  <div
    className="sc-gzVnrw iIRnHj"
  />
  <ul
    className="sc-EHOje kAXgCi pagination"
  >
    <li
      className="page-item disabled"
    >
      <span
        className="page-link"
        disabled={true}
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          className="sr-only"
        >
          Previous
        </span>
      </span>
    </li>
    <li
      className="page-item disabled"
    >
      <span
        className="page-link"
        disabled={true}
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          className="sr-only"
        >
          Next
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`Button component Matches the snapshot 1`] = `
<div
  className="albumsPage"
>
  <div>
    <nav
      className="navbar navbar-expand navbar-dark bg-dark"
    >
      <a
        className="sc-bdVaJa heUcHr navbar-brand"
        href="#home"
      >
        <img
          alt="logo"
          src="logo.svg"
        />
      </a>
      <div
        className="sc-bwzfXH eocOaE"
      >
        React Photo App
      </div>
    </nav>
    <div
      className="sc-htpNat jmLCxE"
    >
      <div
        className="sc-ifAKCX itiWWq"
      >
        Albums
      </div>
    </div>
  </div>
  <div
    className="sc-gzVnrw iIRnHj"
  />
  <ul
    className="sc-EHOje kAXgCi pagination"
  >
    <li
      className="page-item disabled"
    >
      <span
        className="page-link"
        disabled={true}
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ‹
        </span>
        <span
          className="sr-only"
        >
          Previous
        </span>
      </span>
    </li>
    <li
      className="page-item disabled"
    >
      <span
        className="page-link"
        disabled={true}
        onClick={[Function]}
      >
        <span
          aria-hidden="true"
        >
          ›
        </span>
        <span
          className="sr-only"
        >
          Next
        </span>
      </span>
    </li>
  </ul>
</div>
`;
